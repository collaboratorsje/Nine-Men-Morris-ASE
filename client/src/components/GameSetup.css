/* Center the form on the screen */
.game-setup-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center; /* Center content vertically */
    height: 100vh; /* Full viewport height */
    padding: 20px;
    border: 2px solid #b5a37e; /* Matching border color */
    border-radius: 10px; /* Rounded corners */
    background-color: #ffffff; /* White background for the form */
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2); /* Subtle shadow for depth */
    max-width: 600px; /* Restrict maximum width */
    margin: 0 auto; /* Center form horizontally */
}
/* Styling for the header */
.game-setup-header {
    text-align: center;
    width: 100%;
    padding: 10px;
    background-color: #778899; /* Light Steel Gray */
    color: #ffffff; /* White text */
    font-size: 28px;
    font-weight: bold;
    box-shadow: 0px 2px 6px rgba(0, 0, 0, 0.5);
    margin-bottom: 20px; /* Add spacing below header */
}
/* Subheader styling */
.game-setup-container h2,
.game-setup-container h3 {
    font-size: 20px;
    color: #444; /* Dark gray for contrast */
    margin-bottom: 15px;
    text-align: center;
}
/* Input and select styling */
.game-setup-input,
.game-setup-select,
.game-setup-file-input {
    width: 100%;
    max-width: 400px; /* Restrict maximum width */
    padding: 10px;
    margin: 10px 0;
    border: 1px solid #ccc; /* Subtle border for inputs */
    border-radius: 5px; /* Rounded corners for inputs */
    font-size: 16px;
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
    background-color: #f9f9f9; /* Light background for inputs */
}
/* Button styling */
.game-setup-button {
    padding: 10px 20px;
    margin: 10px 5px;
    font-size: 16px;
    border: none;
    border-radius: 5px;
    background-color: #778899; /* Light Steel Gray */
    color: #ffffff; /* White text */
    cursor: pointer;
    transition: background-color 0.3s ease;
}
.game-setup-button:hover {
    background-color: #6b7a8c; /* Darker shade for hover effect */
}
.game-setup-button:disabled {
    background-color: #ccc; /* Gray out disabled button */
    cursor: not-allowed;
}
/* Styling for input groups */
.game-setup-item {
    width: 100%;
    max-width: 400px; /* Restrict maximum width */
    margin-bottom: 15px; /* Add spacing between groups */
    text-align: left;
}
/* Label styling */
.game-setup-label {
    font-size: 16px;
    font-weight: bold;
    color: #333; /* Dark text for readability */
    margin-bottom: 5px;
    display: block;
}
/* Align form elements centrally */
.game-setup-form > * {
    width: 100%;
    max-width: 400px;
    text-align: center;
}
/* Background styling to ensure full height */
body {
    margin: 0;
    padding: 0;
    display: flex;
    justify-content: center; /* Center horizontally */
    align-items: center; /* Center vertically */
    height: 100vh; /* Full viewport height */
    background-color: #d3d3d3; /* Light Steel Gray */
    font-family: 'Arial', sans-serif;
}