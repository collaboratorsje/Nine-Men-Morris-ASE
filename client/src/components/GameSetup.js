import React, { useState } from 'react';

const GameSetup = ({ startGame }) => {
    const [firstPlayer, setFirstPlayer] = useState('player1');
    const [opponentType, setOpponentType] = useState('human');
    const [gameRecord, setGameRecord] = useState(null); // State to store the uploaded game record
    const [gameType, setGameType] = useState('9mm');
    const [autoReplay, setAutoReplay] = useState(false); // New state for auto-replay

    const handleSubmit = (event) => {
        event.preventDefault();
        const options = { firstPlayer, opponentType, gameType };
        console.log('Submitting game setup options:', options);

        fetch('/api/setup', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(options)
        })
            .then(res => res.json())
            .then(data => {
                if (data.success) {
                    startGame(options);
                }
            })
            .catch(error => console.error('Error setting up the game:', error));
    };

    const handleFileUpload = (event) => {
        const file = event.target.files[0];
        if (!file) {
            alert('No file selected.');
            return;
        }
        if (file.type !== "application/json") {
            alert('Invalid file format. Please upload a valid JSON file.');
            return;
        }

        const reader = new FileReader();
        reader.onload = (e) => {
            try {
                const parsedRecord = JSON.parse(e.target.result.trim());
                console.log("Parsed Game Record:", parsedRecord);

                if (parsedRecord && Array.isArray(parsedRecord.moves)) {
                    setGameRecord(parsedRecord); // Store the entire record
                    alert('Game record uploaded successfully! Click Replay Game to start.');
                } else {
                    alert('Invalid JSON structure. Ensure it has a "moves" array.');
                }
            } catch (error) {
                console.error("Error Parsing JSON:", error.message);
                alert('Failed to parse JSON. Ensure the file format is valid.');
            }
        };
        reader.readAsText(file);
    };

    const handleReplayGame = () => {
        if (!gameRecord) {
            alert('No game record loaded. Please upload a file first.');
            return;
        }

        startGame({ 
            firstPlayer: 'replay', 
            opponentType: 'replay', 
            gameRecord: gameRecord.moves, 
            autoReplay // Pass autoReplay flag
        });
    };

    return (
        <form onSubmit={handleSubmit}>
            <h2>Game Setup</h2>

            <div>
                <label>Who goes first: </label>
                <select value={firstPlayer} onChange={(e) => setFirstPlayer(e.target.value)}>
                    <option value="player1">Player 1</option>
                    <option value="player2">Player 2</option>
                </select>
            </div>

            <div>
                <label>Opponent: </label>
                <select value={opponentType} onChange={(e) => setOpponentType(e.target.value)}>
                    <option value="human">Human</option>
                    <option value="computer">Computer</option>
                </select>
            </div>

            <div>
                <label>Game Type: </label>
                <select value={gameType} onChange={(e) => setGameType(e.target.value)}>
                    <option value="9mm">9 men's morris</option>
                    <option value="12mm">12 men's morris</option>
                </select>
            </div>

            <button type="submit">Start Game</button>

            <h3>Load Game Record</h3>
            <input type="file" accept=".json" onChange={handleFileUpload} />
            <div>
                <label>
                    <input 
                        type="checkbox" 
                        checked={autoReplay} 
                        onChange={(e) => setAutoReplay(e.target.checked)} 
                    />
                    Enable Auto Replay
                </label>
            </div>
            <button type="button" onClick={handleReplayGame} disabled={!gameRecord}>
                Replay Game
            </button>
        </form>
    );
};

export default GameSetup;