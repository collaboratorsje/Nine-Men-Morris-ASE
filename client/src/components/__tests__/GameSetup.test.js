import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import GameSetup from '../GameSetup'; // Adjust the import path as necessary

afterEach(() => {
    cleanup();
});

// Test to check if the GameSetup component renders with default values
test('renders GameSetup component with default values', () => {
    render(<GameSetup startGame={jest.fn()} />);

    // Check that the default selections are correct
    expect(screen.getByLabelText(/who goes first/i)).toHaveValue('player1'); // Assuming 'player1' is the default
    expect(screen.getByLabelText(/opponent/i)).toHaveValue('human'); // Assuming 'human' is the default
});

// Test to check that submitting the form calls the startGame function
test('submits the game setup options', () => {
    const startGameMock = jest.fn(); // Mock the startGame function
    render(<GameSetup startGame={startGameMock} />);

    // Check the default selections
    expect(screen.getByLabelText(/who goes first/i)).toHaveValue({firstPlayer});
    expect(screen.getByLabelText(/opponent/i)).toHaveValue('human');
    
    // Change selections
    fireEvent.change(screen.getByLabelText(/who goes first/i), { target: { value: 'player2' } });
    fireEvent.change(screen.getByLabelText(/opponent/i), { target: { value: 'computer' } });
    
    // Verify the updated selections
    expect(screen.getByLabelText(/who goes first/i)).toHaveValue('player2');
    expect(screen.getByLabelText(/opponent/i)).toHaveValue('computer');
    // Submit the form
    fireEvent.click(screen.getByText(/start game/i));

    // Check that startGame was called with the correct options
    expect(startGameMock).toHaveBeenCalledWith({ firstPlayer: 'player2', opponentType: 'computer' });
});

// Test for snapshot
test('matches snapshot', () => {
    const tree = renderer.create(<GameSetup startGame={jest.fn()} />).toJSON();
    expect(tree).toMatchSnapshot();
});