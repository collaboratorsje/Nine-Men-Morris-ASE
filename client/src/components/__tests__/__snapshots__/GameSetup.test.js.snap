// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Game Setup
  </h2>
  <div>
    <label>
      Who goes first: 
    </label>
    <select
      onChange={[Function]}
      value="player1"
    >
      <option
        value="player1"
      >
        Player 1
      </option>
      <option
        value="player2"
      >
        Player 2
      </option>
    </select>
  </div>
  <div>
    <label>
      Opponent: 
    </label>
    <select
      onChange={[Function]}
      value="human"
    >
      <option
        value="human"
      >
        Human
      </option>
      <option
        value="computer"
      >
        Computer
      </option>
    </select>
  </div>
  <button
    type="submit"
  >
    Start Game
  </button>
</form>
`;
